<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
  DangerID: danger-id-default;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.5.1 test /Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10726
> grunt

Running "appcJs:src:lintOnly" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/ium-sdk_titanium_mobile_PR-10726/tests/Resources/ti.media.audioplayer.addontest.js
  29:55  warning  'e' is defined but never used  no-unused-vars

✖ 1 problem (0 errors, 1 warning)


Running "checkFormat:ios" (checkFormat) task

Running "checkFormat:android" (checkFormat) task
Fatal error: Error: Formatting incorrect on "android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java", proposed changes: <?xml version='1.0'?>
<replacements xml:space='preserve' incomplete_format='false'>
<replacement offset='3282' length='6'>&#10;&#10;		</replacement>
</replacements>

npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

--- a/android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java
+++ b/android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java
@@ -12,12 +12,10 @@ import org.appcelerator.kroll.KrollModule;
 import org.appcelerator.kroll.annotations.Kroll;
 import org.appcelerator.kroll.common.Log;
 import org.appcelerator.titanium.TiApplication;
+import org.appcelerator.titanium.util.TiLocaleManager;
 import org.appcelerator.titanium.util.TiPlatformHelper;
 import org.appcelerator.titanium.util.TiRHelper;
 
-import android.content.Context;
-import android.content.res.Configuration;
-import android.os.Build;
 import android.telephony.PhoneNumberUtils;
 
 @Kroll.module
@@ -106,17 +104,8 @@ public class LocaleModule extends KrollModule
 				locale = new Locale(parts[0]);
 			}
 
-			Locale.setDefault(locale);
+			TiLocaleManager.setLocale(locale);
 
-			Configuration config = new Configuration();
-			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
-				config.setLocale(locale);
-			} else {
-				config.locale = locale;
-			}
-
-			Context ctx = TiApplication.getAppRootOrCurrentActivity().getBaseContext();
-			ctx.getResources().updateConfiguration(config, ctx.getResources().getDisplayMetrics());
 		} catch (Exception e) {
 			Log.e(TAG, "Error trying to set language '" + language + "':", e);
 		}

--- a/android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java
+++ b/android/modules/locale/src/java/ti/modules/titanium/locale/LocaleModule.java
@@ -131,7 +131,7 @@ public class LocaleModule extends KrollModule
 		if (defaultValue == null) {
 			defaultValue = key;
 		}
-		
+
 		try {
 			int resid = TiRHelper.getResource("string." + key.replace(".", "_"));
 			if (resid != 0) {

--- a/android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/TiBaseActivity.java
@@ -37,6 +37,7 @@ import org.appcelerator.titanium.util.TiActivityResultHandler;
 import org.appcelerator.titanium.util.TiActivitySupport;
 import org.appcelerator.titanium.util.TiActivitySupportHelper;
 import org.appcelerator.titanium.util.TiConvert;
+import org.appcelerator.titanium.util.TiLocaleManager;
 import org.appcelerator.titanium.util.TiMenuSupport;
 import org.appcelerator.titanium.util.TiPlatformHelper;
 import org.appcelerator.titanium.util.TiUIHelper;
@@ -645,6 +646,12 @@ public abstract class TiBaseActivity extends AppCompatActivity implements TiActi
 	{
 		overridenLayout = true;
 		super.setContentView(view, params);
+	}
+
+	@Override
+	protected void attachBaseContext(Context newBase)
+	{
+		super.attachBaseContext(TiLocaleManager.getLocalizedContext(newBase));
 	}
 
 	@Override

--- a/android/titanium/src/java/org/appcelerator/titanium/proxy/ActivityProxy.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/proxy/ActivityProxy.java
@@ -87,6 +87,14 @@ public class ActivityProxy extends KrollProxy implements TiActivityResultHandler
 	}
 
 	@Kroll.method
+	public void recreate()
+	{
+		if (activity.get() != null) {
+			activity.get().recreate();
+		}
+	}
+
+	@Kroll.method
 	public DecorViewProxy getDecorView()
 	{
 		if (savedDecorViewProxy == null) {

--- a/android/titanium/src/java/org/appcelerator/titanium/util/TiLocaleManager.java
+++ b/android/titanium/src/java/org/appcelerator/titanium/util/TiLocaleManager.java
@@ -1,0 +1,39 @@
+/**
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2009-present by Axway, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+package org.appcelerator.titanium.util;
+
+import android.content.Context;
+import android.content.res.Configuration;
+
+import org.appcelerator.titanium.TiApplication;
+
+import java.util.Locale;
+
+public class TiLocaleManager
+{
+
+	private static Locale currentLocale = Locale.getDefault();
+
+	public static Context getLocalizedContext(Context ctx)
+	{
+		Configuration config = ctx.getResources().getConfiguration();
+		config.setLocale(currentLocale);
+		return ctx.createConfigurationContext(config);
+	}
+
+	public static void setLocale(Locale locale)
+	{
+		Locale.setDefault(locale);
+		// Resorting to the deprecated method updateConfiguration in order to immediately change
+		// the localized strings in the Activity that has triggered the locale change.
+		Context context = TiApplication.getInstance().getBaseContext();
+		Configuration configuration = new Configuration();
+		configuration.setLocale(locale);
+		context.getResources().updateConfiguration(configuration, context.getResources().getDisplayMetrics());
+		currentLocale = locale;
+	}
+}

--- a/apidoc/Titanium/Android/Activity.yml
+++ b/apidoc/Titanium/Android/Activity.yml
@@ -258,6 +258,11 @@ methods:
         summary: Name of the permission that the receiver should hold in order to receive the broadcast.
         type: String
         optional: true
+
+  - name: recreate
+    summary: Recreates the activity.
+    since: "8.0.0"
+
 events:
 
   - name: create

--- a/apidoc/Titanium/Locale/Locale.yml
+++ b/apidoc/Titanium/Locale/Locale.yml
@@ -71,6 +71,10 @@ methods:
         [ISO 3166-1 alpha-2](http://en.wikipedia.org/wiki/ISO_3166-1_alpha-2#Officially_assigned_code_elements) 
         sections of wikipedia for reference.
 
+        In case the new language is with different layout direction the activity from where the change is called
+        and all activities currently opened will need to recreated in order to be properly shown. You can do that
+        with [recreate](Ti.Android.Activity.recreate).
+
     platforms: [iphone, ipad, android]
     parameters: 
       - name: language

--- a/tests/Resources/ti.locale.test.js
+++ b/tests/Resources/ti.locale.test.js
@@ -1,0 +1,177 @@
+/*
+ * Appcelerator Titanium Mobile
+ * Copyright (c) 2015-Present by Appcelerator, Inc. All Rights Reserved.
+ * Licensed under the terms of the Apache Public License
+ * Please see the LICENSE included with this distribution for details.
+ */
+/* eslint-env mocha */
+/* global Ti, L */
+/* eslint no-unused-expressions: "off" */
+'use strict';
+var should = require('./utilities/assertions');
+
+describe('Global', function () {
+	it('L', function () {
+		should(L).be.a.Function;
+		// should(L).eql(Ti.Locale.getString);
+	});
+});
+
+describe('Titanium.Locale', function () {
+
+	// reset back to US english when done
+	after(function () {
+		Ti.Locale.setLanguage('en-US');
+	});
+
+	it('apiName', function () {
+		should(Ti.Locale).have.a.readOnlyProperty('apiName').which.is.a.String;
+		should(Ti.Locale.apiName).be.eql('Ti.Locale');
+	});
+
+	it('exists', function () {
+		should(Ti.Locale).not.be.undefined;
+		should(Ti.Locale).not.be.null;
+		should(Ti.Locale).be.an.Object;
+	});
+
+	it('#getString()', function () {
+		should(Ti.Locale.getString).be.a.Function;
+	});
+
+	it('#getCurrentCountry()', function () {
+		should(Ti.Locale.getCurrentCountry).be.a.Function;
+		should(Ti.Locale.getCurrentCountry()).eql('US');
+	});
+
+	it('#getCurrentLanguage()', function () {
+		should(Ti.Locale.getCurrentLanguage).be.a.Function;
+		should(Ti.Locale.getCurrentLanguage()).eql('en');
+	});
+
+	it('#getLocaleCurrencySymbol', function () {
+		should(Ti.Locale.getLocaleCurrencySymbol).be.a.Function;
+		should(Ti.Locale.getLocaleCurrencySymbol('en-US')).eql('$');
+	});
+
+	// FIXME Get working on iOS
+	// FIXME Get working properly cross-platform. JPY gives us ¥ on Windows and Android, JP¥ on iOS. CNY gives us ¥ on Windows, CN¥ on Android
+	it.androidAndIosBroken('#getCurrencySymbol(String)', function () {
+		should(Ti.Locale.getCurrencySymbol).be.a.Function;
+		should(Ti.Locale.getCurrencySymbol('USD')).eql('$');
+		should(Ti.Locale.getCurrencySymbol('JPY')).eql('¥'); // 'JP¥' on iOS
+		should(Ti.Locale.getCurrencySymbol('CNY')).eql('¥'); // 'CN¥' on Android
+		should(Ti.Locale.getCurrencySymbol('TWD')).eql('NT$');
+	});
+
+	it('#getCurrencyCode(String)', function () {
+		should(Ti.Locale.getCurrencyCode).be.a.Function;
+		should(Ti.Locale.getCurrencyCode('en-US')).eql('USD');
+		should(Ti.Locale.getCurrencyCode('ja-JP')).eql('JPY');
+		should(Ti.Locale.getCurrencyCode('zh-CN')).eql('CNY');
+		should(Ti.Locale.getCurrencyCode('zh-TW')).eql('TWD');
+	});
+
+	// TODO Support Ti.Locale.formatTelephoneNumber on other platforms?
+	it.android('#formatTelephoneNumber(String)', function () {
+		should(Ti.Locale.formatTelephoneNumber).be.a.Function;
+		// TODO Actually check inputs/outputs!
+	});
+
+	it('currentCountry', function () {
+		should(Ti.Locale.currentCountry).be.a.String;
+		should(Ti.Locale.currentCountry).eql('US');
+	});
+
+	it('currentLanguage', function () {
+		should(Ti.Locale.currentLanguage).be.a.String;
+		should(Ti.Locale.currentLanguage).eql('en');
+	});
+
+	it('currentLocale', function () {
+		should(Ti.Locale.currentLocale).be.a.String;
+		should(Ti.Locale.currentLocale).eql('en-US');
+	});
+
+	describe('#setLanguage(String)', function () {
+		it('is a Function', function () {
+			should(Ti.Locale.setLanguage).be.a.Function;
+		});
+
+		it('changes currentLanguage', function () {
+			Ti.Locale.setLanguage('fr');
+			should(Ti.Locale.currentLanguage).eql('fr');
+		});
+
+		// FIXME Get working on iOS, setLangauge doesn't seem to affect currentLocale
+		it.iosBroken('changes currentLocale', function () {
+			Ti.Locale.setLanguage('en-GB');
+			should(Ti.Locale.currentLocale).eql('en-GB'); // iOS returns 'en-US'
+			Ti.Locale.setLanguage('fr');
+			should(Ti.Locale.currentLocale).eql('fr');
+		});
+
+		// TODO test if it changes the currentCountry?
+	});
+
+	describe('#getString()', function () {
+		it('is a Function', function () {
+			should(Ti.Locale.getString).be.a.Function;
+		});
+
+		beforeEach(function () {
+			Ti.Locale.setLanguage('en-US');
+		});
+
+		it('returns stored value for found key', function () {
+			should(Ti.Locale.getString('this_is_my_key')).eql('this is my value');
+			should(L('this_is_my_key')).eql('this is my value');
+		});
+
+		it('returns key if not found and no default specified', function () {
+			should(Ti.Locale.getString('this_should_not_be_found')).eql('this_should_not_be_found');
+			should(L('this_should_not_be_found')).eql('this_should_not_be_found');
+		});
+
+		it('returns supplied default if key not found', function () {
+			should(Ti.Locale.getString('this_should_not_be_found', 'this is the default value')).eql('this is the default value');
+			should(L('this_should_not_be_found', 'this is the default value')).eql('this is the default value');
+		});
+
+		it('returns key if supplied default is not a String and key/value pair not found', function () {
+			should(Ti.Locale.getString('this_should_not_be_found', null)).eql('this_should_not_be_found');
+			should(L('this_should_not_be_found', null)).eql('this_should_not_be_found');
+			should(Ti.Locale.getString('this_should_not_be_found', 123)).eql('this_should_not_be_found');
+			should(L('this_should_not_be_found', 123)).eql('this_should_not_be_found');
+		});
+
+		it('handles locale/country specific languages (i.e. en-GB vs en-US)', function () {
+			Ti.Locale.setLanguage('en-GB');
+			should(Ti.Locale.getString('this_is_my_key')).eql('this is my en-GB value'); // This fails on Windows, gives 'this is my value'
+			should(L('this_is_my_key')).eql('this is my en-GB value'); // This fails on Windows, gives 'this is my value'
+		});
+
+		it('handles single segment language (i.e. ja)', function () {
+			Ti.Locale.setLanguage('ja');
+			should(Ti.Locale.getString('this_is_my_key')).eql('これは私の値です');
+			should(L('this_is_my_key')).eql('これは私の値です');
+		});
+
+		// FIXME iOS seems to ignore position info on the format string.
+		// We're trying to force the 1st argument into the second slot, and vice versa here. iOS handles the %2$s syntax, but ignores position
+		it.iosBroken('usage with String.format()', function () {
+			var i18nMissingMsg = '<no translation available>';
+			var string1 = 'You say ' + Ti.Locale.getString('signoff', i18nMissingMsg) + ' and I say ' + Ti.Locale.getString('greeting', i18nMissingMsg) + '!';
+			var string2 = String.format(L('phrase'), L('greeting', i18nMissingMsg), L('signoff', i18nMissingMsg));
+
+			should(string1).eql(string2);
+			if (Ti.Locale.currentLanguage === 'en') {
+				should(string1).eql('You say goodbye and I say hello!');
+				should(string2).eql('You say goodbye and I say hello!');
+			} else if (Ti.Locale.currentLanguage === 'ja') {
+				should(string1).eql('You say さようなら and I say こんにちは!');
+				should(string2).eql('You say さようなら and I say こんにちは!');
+			}
+		});
+	});
+});

